# Data Science Pandas

## Tasks Today:

0) <b>Pre-Work</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Numpy Random Sampling

1) <b>Pandas</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Importing <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Tabular Data Structures <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - from_dict() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - read_csv() <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) <b>In-Class Exercise #1</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Accessing Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Indexing <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - df.loc <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - keys() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Slicing a DataFrame <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Built-In Methods <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - head() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - tail() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - shape <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - describe() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - sort_values() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - .columns <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) <b>In-Class Exercise #2</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; g) Filtration <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Conditionals <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Subsetting <br>
 &nbsp;&nbsp;&nbsp;&nbsp; h) Column Transformations <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Generating a New Column w/Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - User Defined Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; i) Aggregations <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - groupby() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Type of groupby() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - mean() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - groupby() w/Multiple Columns <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - drop_duplicates() <br>

## Numpy Random Sampling

In [1]:
import numpy as np

# A single call generates a single random number between 0 and 1

print('Here is a random number: %s' % np.random.uniform())

# You can also pass some parameters or bounds

print('Here is a random number between 0 and 1 Million: %s' % np.random.uniform(0, 1e6))

# You can also generate a bunch of random numbers all at once

print('Here are 3 random numbers between 0 and 10: %s' % np.random.uniform(0, 10, 3))

# Even matrices with shapes as a parameter

print('Here is a 3x3 matrix with numbers between 0 and 10: \n %s' % np.random.uniform(0, 10, (3, 3)))


Here is a random number: 0.9643928418215584
Here is a random number between 0 and 1 Million: 932917.7272920033
Here are 3 random numbers between 0 and 10: [0.40720219 5.69620844 7.25899206]
Here is a 3x3 matrix with numbers between 0 and 10: 
 [[6.15092423 7.3765736  0.10976245]
 [6.58972247 0.40883183 9.85102628]
 [1.9391081  6.76102246 5.6734346 ]]


## Pandas <br>

<p>Pandas is a flexible data analysis library built on top of NumPy that is excellent for working with tabular data. It is currently the de-facto standard for Python-based data analysis, and fluency in Pandas will do wonders for your productivity and frankly your resume. It is one of the fastest ways of getting from zero to answer in existence. </p>

<ul>
    <li>Pandas is a Python module, written in C. The Pandas module is a high performance, highly efficient, and high level data analysis library. It allows us to work with large sets of data called dataframes.</li>
    <li>Series is a one-dimensional labeled array capable of holding data of any type (integer, string, float, python objects, etc.)</li>
    <li>Dataframe = Spreadsheet (has column headers, index, etc.)</li>
</ul>

### Importing

In [2]:
!pip install pandas

zsh:1: command not found: pip


In [3]:
import pandas as pd

# always use pd, standard for data science

### Tabular data structures <br>
<p>The central object of study in Pandas is the DataFrame, which is a tabular data structure with rows and columns like an excel spreadsheet. The first point of discussion is the creation of dataframes both from native Python dictionaries, and text files through the Pandas I/O system.</p>

In [4]:
names = ['Alice',
         'Bob',
         'James',
         'Beth',
         'John',
         'Sally',
         'Richard',
         'Lauren',
         'Brandon',
         'Sabrina']

ages = np.random.randint(10, 35, len(names))

my_people = {
    'names' : names,
    'ages' : ages
}

my_people

{'names': ['Alice',
  'Bob',
  'James',
  'Beth',
  'John',
  'Sally',
  'Richard',
  'Lauren',
  'Brandon',
  'Sabrina'],
 'ages': array([25, 14, 13, 17, 29, 24, 22, 28, 13, 14])}

##### from_dict()

<p>Let's convert our not-so-useful-for-analysis dict into a Pandas dataframe. We can use the from_dict function to do this easily using Pandas:</p>

In [5]:
data = pd.DataFrame.from_dict(my_people)

data

names  ages
0    Alice    25
1      Bob    14
2    James    13
3     Beth    17
4     John    29
5    Sally    24
6  Richard    22
7   Lauren    28
8  Brandon    13
9  Sabrina    14

##### read_csv()

In [6]:
# pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', None)

marathon = pd.read_csv('boston_marathon2017_edited.csv', sep = ',')
marathon

10K                     15K                    20K  \
0      12/30/1899 12:30:28 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
1      12/30/1899 12:30:27 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
2      12/30/1899 12:30:29 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:16 AM   
3      12/30/1899 12:30:29 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:19 AM   
4      12/30/1899 12:30:28 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
...                       ...                     ...                    ...   
26405   12/30/1899 1:35:41 AM   12/30/1899 2:23:35 AM  12/30/1899 3:12:44 AM   
26406   12/30/1899 1:05:33 AM   12/30/1899 1:52:17 AM  12/30/1899 2:49:41 AM   
26407   12/30/1899 1:43:36 AM   12/30/1899 2:32:36 AM                    NaN   
26408   12/30/1899 1:27:19 AM   12/30/1899 2:17:17 AM  12/30/1899 3:11:40 AM   
26409   12/30/1899 1:17:12 AM   12/30/1899 2:00:10 AM  12/30/1899 2:58:55 AM   

                         25K                    30K                    35K  \
0      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:19 AM   
1      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:19 AM   
2      12/30/1899 1:17:00 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:31 AM   
3      12/30/1899 1:17:00 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:58 AM   
4      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:41 AM   
...                      ...                    ...                    ...   
26405  12/30/1899 4:12:06 AM  12/30/1899 5:03:08 AM  12/30/1899 5:55:18 AM   
26406  12/30/1899 3:50:19 AM  12/30/1899 4:50:01 AM  12/30/1899 5:53:48 AM   
26407  12/30/1899 4:15:21 AM  12/30/1899 5:06:37 AM  12/30/1899 6:00:33 AM   
26408  12/30/1899 4:06:10 AM  12/30/1899 5:07:09 AM  12/30/1899 6:06:07 AM   
26409  12/30/1899 4:27:14 AM  12/30/1899 5:37:13 AM  12/30/1899 6:39:07 AM   

                         40K                      5K  Age      Bib  ...  \
0      12/30/1899 2:02:53 AM  12/30/1899 12:15:25 AM   24     11.0  ...   
1      12/30/1899 2:03:14 AM  12/30/1899 12:15:24 AM   30     17.0  ...   
2      12/30/1899 2:03:38 AM  12/30/1899 12:15:25 AM   25     23.0  ...   
3      12/30/1899 2:04:35 AM  12/30/1899 12:15:25 AM   32     21.0  ...   
4      12/30/1899 2:05:00 AM  12/30/1899 12:15:25 AM   31      9.0  ...   
...                      ...                     ...  ...      ...  ...   
26405  12/30/1899 6:46:57 AM  12/30/1899 12:46:44 AM   61  25166.0  ...   
26406  12/30/1899 6:54:21 AM  12/30/1899 12:32:03 AM   25  25178.0  ...   
26407  12/30/1899 6:54:38 AM  12/30/1899 12:53:11 AM   57  27086.0  ...   
26408  12/30/1899 6:56:08 AM  12/30/1899 12:40:34 AM   64  25268.0  ...   
26409  12/30/1899 7:41:23 AM  12/30/1899 12:39:36 AM   48  25266.0  ...   

       Division Gender                   Half  M/F                      Name  \
0             1      1  12/30/1899 1:04:35 AM    M           Kirui, Geoffrey   
1             2      2  12/30/1899 1:04:35 AM    M               Rupp, Galen   
2             3      3  12/30/1899 1:04:36 AM    M             Osako, Suguru   
3             4      4  12/30/1899 1:04:45 AM    M          Biwott, Shadrack   
4             5      5  12/30/1899 1:04:35 AM    M            Chebet, Wilson   
...         ...    ...                    ...  ...                       ...   
26405       344  11972  12/30/1899 3:23:31 AM    F  Steinbach, Paula Eyvonne   
26406      4774  14436  12/30/1899 3:00:26 AM    M        Avelino, Andrew R.   
26407       698  11973  12/30/1899 3:36:24 AM    F           Hantel, Johanna   
26408      1043  14437  12/30/1899 3:22:30 AM    M              Reilly, Bill   
26409      2553  14438  12/30/1899 3:08:16 AM    M             Rigsby, Scott   

      Number of Records          Official Time  Overall  \
0                     1  12/30/1899 2:09:37 AM        1   
1                     1  12/30/1899 2:09:58 AM        2   
2                     1  12/30/1899 2:10:28 AM        3   
3                     1  12/30

### In-Class Exercise #1 - Read in Boston Red Sox Hitting Data <br>
<p>Use the pandas read_csv() method to read in the statistics from the two files yesterday.</p>

In [10]:
bos2017 = pd.read_csv('redsox_2017_hitting.txt', sep = ',')
bos2018 = pd.read_csv('redsox_2018_hitting.txt', sep = ',')

bos2017
bos2018

Rk Pos                Name  Age    G   PA   AB    R    H  2B  ...    OBP  \
0    1   C          Sandy Leon   29   89  288  265   30   47  12  ...  0.232   
1    2  1B      Mitch Moreland   32  124  459  404   57   99  23  ...  0.325   
2    3  2B       Eduardo Nunez   31  127  502  480   56  127  23  ...  0.289   
3    4  SS     Xander Bogaerts   25  136  580  513   72  148  45  ...  0.360   
4    5  3B       Rafael Devers   21  121  490  450   59  108  24  ...  0.298   
5    6  LF   Andrew Benintendi   23  148  661  579  103  168  41  ...  0.366   
6    7  CF  Jackie Bradley Jr.   28  144  535  474   76  111  33  ...  0.314   
7    8  RF        Mookie Betts   25  136  614  520  129  180  47  ...  0.438   
8    9  DH       J.D. Martinez   30  150  649  569  111  188  37  ...  0.402   
9   10  MI          Brock Holt   30  109  367  321   41   89  18  ...  0.362   
10  11   C   Christian Vazquez   27   80  269  251   24   52  10  ...  0.257   
11  12  UT       Blake Swihart   26   82  207  192   28   44  10  ...  0.285   
12  13  UT      Hanley Ramirez   34   44  195  177   25   45   7  ...  0.313   
13  14  1B        Steve Pearce   35   50  165  136   19   38   8  ...  0.394   
14  15  2B         Ian Kinsler   36   37  143  132   17   32   6  ...  0.294   
15  16  SS         Tzu-Wei Lin   24   37   73   65   15   16   6  ...  0.329   
16  17  UT          Sam Travis   24   19   38   36    5    8   3  ...  0.263   
17  18  UT    Brandon Phillips   37    9   27   23    4    3   0  ...  0.259   
18  19  2B      Dustin Pedroia   34    3   13   11    1    1   0  ...  0.231   
19  20   C          Dan Butler   31    2    7    6    0    1   0  ...  0.143   

      SLG    OPS  OPS+   TB  GDP  HBP  SH  SF  IBB  
0   0.279  0.511    37   74    6    4   3   1    0  
1   0.433  0.758   102  175   12    0   0   5    2  
2   0.388  0.677    81  186   17    2   1   3    0  
3   0.522  0.883   135  268   14    6   0   6    4  
4   0.433  0.731    94  195    9    0   0   2    6  
5   0.465  0.830   123  269    9    2   2   7    1  
6   0.403  0.717    92  191    6   11   0   4    3  
7   0.640  1.078   186  333    5    8   0   5    8  
8   0.629  1.031   173  358   19    4   0   7   11  
9   0.411  0.774   109  132    7    7   0   2    2  
10  0.283  0.540    46   71    5    4   1   0    1  
11  0.328  0.613    65   63    4    0   0   0    0  
12  0.395  0.708    90   70    9    2   0   2    0  
13  0.507  0.901   141   69    1    5   0   2    0  
14  0.311  0.604    64   41    5    0   0   1    0  
15  0.415  0.744   100   27    0    0   0   0    0  
16  0.389  0.652    73   14    1    0   0   0    0  
17  0.261  0.520    42    6    1    0   0   0    0  
18  0.091  0.322    -7    1    0    0   0   0    0  
19  0.167  0.310   -17    1    0    0   0   1    0  

[20 rows x 28 columns]

### Accessing Data <br>

##### Indexing

<p>You can directly select a column of a dataframe just like you would a dict. The result is a Pandas 'Series' object.</p>

In [15]:
data_ages = data['ages']

print(data_ages)

# print out the type of the variable data
# Series
print(type(data_ages))

# DataFrame
print(type(data))

# DataSeries Index - starting at 0
print(data_ages[3])

# key first, row index second
print(data['ages'][3])

# change the data type for value
print(float(data_ages[0]))

0    25
1    14
2    13
3    17
4    29
5    24
6    22
7    28
8    13
9    14
Name: ages, dtype: int64
<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>
17
17
25.0


##### df.loc

<p>Along the horizontal dimension, rows of Pandas DataFrames are Row objects. You will notice there is a third column present in the DataFrame - this is the $\textit{index}$. It is automatically generated as a row number, but can be reassigned to a column of your choice using the DataFrame.set_index(colname) method. We can use it to access particular Pandas $\textit{rows}$, which are also Series objects:</p>

In [126]:
print(data)

# Grab the first row of data -- by the index of the row
my_row = data.loc[3]
print(f'\n Location of the 3rd row in the data: \n{my_row}')

# Getting Multiple values from a df.loc
multiple_peeps = data.loc[  [0, 1, 2]  ]['names']
print(f'\n Location of multiple people: {multiple_peeps}')

# Using df.loc and producing a DataFrame at the end
multiple_peeps_df = data.loc[[0, 1, 2]][['names', 'ages']]
print('\n', multiple_peeps_df)
print('\n', type(multiple_peeps_df))
# Reset our my_people dataframe index
reset_index = data.loc[[0, 1, 2, 3]].set_index('names')
print(reset_index.loc['Alice'])


     names  ages
0    Alice    25
1      Bob    14
2    James    13
3     Beth    17
4     John    29
5    Sally    24
6  Richard    22
7   Lauren    28
8  Brandon    13
9  Sabrina    14

 Location of the 3rd row in the data: 
names    Beth
ages       17
Name: 3, dtype: object

 Location of multiple people: 0    Alice
1      Bob
2    James
Name: names, dtype: object

    names  ages
0  Alice    25
1    Bob    14
2  James    13

 <class 'pandas.core.frame.DataFrame'>
ages    25
Name: Alice, dtype: int64


##### keys()

In [30]:
# Access all of the keys/columns of the dataframe
# Dataframe.keys()

data.keys()
print(data.keys())
print(type(bos2017.keys()))
print(type(bos2017.keys().tolist()))
print(bos2017.keys())


Index(['names', 'ages'], dtype='object')
<class 'pandas.core.indexes.base.Index'>
<class 'list'>
Index(['Rk', 'Pos', 'Name', 'Age', 'G', 'PA', 'AB', 'R', 'H', '2B', '3B', 'HR',
       'RBI', 'SB', 'CS', 'BB', 'SO', 'BA', 'OBP', 'SLG', 'OPS', 'OPS+', 'TB',
       'GDP', 'HBP', 'SH', 'SF', 'IBB'],
      dtype='object')


##### Slicing a DataFrame

In [38]:
# printing all data for context
print(data)

data[2:5]

     names  ages
0    Alice    25
1      Bob    14
2    James    13
3     Beth    17
4     John    29
5    Sally    24
6  Richard    22
7   Lauren    28
8  Brandon    13
9  Sabrina    14


names  ages
2  James    13
3   Beth    17
4   John    29

### Built-In Methods <br>

<p>These are methods that are frequently used when using Pandas to make your life easier. It is possible to spend a whole week simply exploring the built-in functions supported by DataFrames in Pandas. Here however, we will simply highlight a few ones that might be useful, to give you an idea of what's possible out of the box with Pandas:</p>

##### head()

In [39]:
# DataFrame.head()  -- Accepts integer parameter(gives access to more rows)

bos2017.head()

Rk Pos               Name  Age    G   PA   AB   R    H  2B  ...    OBP  \
0   1   C  Christian Vazquez   26   99  345  324  43   94  18  ...  0.330   
1   2  1B     Mitch Moreland   31  149  576  508  73  125  34  ...  0.326   
2   3  2B     Dustin Pedroia   33  105  463  406  46  119  19  ...  0.369   
3   4  SS    Xander Bogaerts   24  148  635  571  94  156  32  ...  0.343   
4   5  3B      Rafael Devers   20   58  240  222  34   63  14  ...  0.338   

     SLG    OPS  OPS+   TB  GDP  HBP  SH  SF  IBB  
0  0.404  0.735    91  131   14    3   0   1    0  
1  0.443  0.769    99  225   14    6   0   5    6  
2  0.392  0.760   100  159   11    2   2   4    4  
3  0.403  0.746    95  230   17    6   0   2    6  
4  0.482  0.819   111  107    5    0   0   0    3  

[5 rows x 28 columns]

##### tail()

In [40]:
# DataFrame.tail()  -- Accepts integer parameter(gives access to more rows)

bos2017.tail()

Rk Pos             Name  Age   G  PA  AB  R   H  2B  ...    OBP    SLG  \
18  19  IF  Marco Hernandez   24  21  60  58  7  16   3  ...  0.300  0.328   
19  20  UT      Rajai Davis   36  17  38  36  7   9   2  ...  0.289  0.306   
20  21  UT     Steve Selsky   27   8   9   9  0   1   1  ...  0.111  0.222   
21  22  UT    Blake Swihart   25   6   7   5  1   1   0  ...  0.429  0.200   
22  23  2B   Chase d'Arnaud   30   2   1   1  2   1   0  ...  1.000  1.000   

      OPS  OPS+  TB  GDP  HBP  SH  SF  IBB  
18  0.628    65  19    0    1   0   0    0  
19  0.595    56  11    2    1   0   0    0  
20  0.333   -16   2    0    0   0   0    0  
21  0.629    74   1    0    0   0   0    0  
22  2.000   428   1    0    0   0   0    0  

[5 rows x 28 columns]

##### shape

In [41]:
# The dataframe has a shape property, just like a NumPy matrix. 
# print(df.shape) -- DataFrame.shape -- No Parameter


print(bos2017.shape)


(23, 28)


##### describe() <br>
<p>Probably one of the most important methods to understand...</p>

In [46]:
# Collect summary statistics in one line
# DataFrame.describe() -- Accepts parameters (include, exclude)

bos2018.describe(include = 'all')


Rk  Pos        Name        Age           G          PA  \
count   20.00000   20          20  20.000000   20.000000   20.000000   
unique       NaN   11          20        NaN         NaN         NaN   
top          NaN   UT  Sandy Leon        NaN         NaN         NaN   
freq         NaN    4           1        NaN         NaN         NaN   
mean    10.50000  NaN         NaN  29.100000   82.350000  314.100000   
std      5.91608  NaN         NaN   4.666792   52.916294  230.972065   
min      1.00000  NaN         NaN  21.000000    2.000000    7.000000   
25%      5.75000  NaN         NaN  25.000000   37.000000  125.500000   
50%     10.50000  NaN         NaN  29.500000   85.500000  278.500000   
75%     15.25000  NaN         NaN  32.500000  129.250000  510.250000   
max     20.00000  NaN         NaN  37.000000  150.000000  661.000000   

                AB           R           H         2B  ...        OBP  \
count    20.000000   20.000000   20.000000  20.000000  ...  20.000000   
unique         NaN         NaN         NaN        NaN  ...        NaN   
top            NaN         NaN         NaN        NaN  ...        NaN   
freq           NaN         NaN         NaN        NaN  ...        NaN   
mean    280.200000   43.600000   75.250000  17.650000  ...   0.307700   
std     203.871631   38.293053   62.462515  15.624795  ...   0.068843   
min       6.000000    0.000000    1.000000   0.000000  ...   0.143000   
25%     115.250000   16.500000   28.000000   6.000000  ...   0.262000   
50%     258.000000   29.000000   49.500000  11.000000  ...   0.305500   
75%     475.500000   62.250000  115.000000  26.250000  ...   0.360500   
max     579.000000  129.000000  188.000000  47.000000  ...   0.438000   

              SLG        OPS        OPS+          TB        GDP        HBP  \
count   20.000000  20.000000   20.000000   20.000000  20.000000  20.000000   
unique        NaN        NaN         NaN         NaN        NaN        NaN   
top           NaN        NaN         NaN         NaN        NaN        NaN   
freq          NaN        NaN         NaN         NaN        NaN        NaN   
mean     0.387500   0.695200   86.450000  127.200000   6.500000   2.750000   
std      0.135884   0.200419   52.199491  113.226183   5.633546   3.274704   
min      0.091000   0.310000  -17.000000    1.000000   0.000000   0.000000   
25%      0.304000   0.588000   59.500000   37.500000   1.000000   0.000000   
50%      0.399000   0.712500   91.000000   72.500000   5.500000   2.000000   
75%      0.441000   0.788000  112.500000  192.000000   9.000000   4.250000   
max      0.640000   1.078000  186.000000  358.000000  19.000000  11.000000   

               SH         SF        IBB  
count   20.000000  20.000000  20.000000  
unique        NaN        NaN        NaN  
top           NaN        NaN        NaN  
freq          NaN        NaN        NaN  
mean     0.350000   2.400000   1.900000  
std      0.812728   2.436564   3.110255  
min      0.000000   0.000000   0.000000  
25%      0.000000   0.000000   0.000000  
50%      0.000000   2.000000   0.000000  
75%      0.000000   4.250000   2.250000  
max      3.000000   7.000000  11.000000  

[11 rows x 28 columns]

##### sort_values()

In [49]:
# Sort based on many labels, with left-to-right priority
# sorted_data = data.sort_values('ages').reset_index()

# DataFrame.sort_values('key')

sorted_data = data.sort_values('ages', kind = 'mergesort')
sorted_data.reset_index(drop = True)



names  ages
0    James    13
1  Brandon    13
2      Bob    14
3  Sabrina    14
4     Beth    17
5  Richard    22
6    Sally    24
7    Alice    25
8   Lauren    28
9     John    29

##### .columns

In [52]:
# will show all cols headers
# DataFrame.columns -- has no parameters

print(bos2017.columns)
print(f'\n The type of bos2017.columns: {type(bos2017.columns)}')


# Keys brings back the 'index' of whatever data type we are working with 
print(bos2017.keys())
print(f'\n The type of bos2017.keys(): {type(bos2017.keys())}')


Index(['Rk', 'Pos', 'Name', 'Age', 'G', 'PA', 'AB', 'R', 'H', '2B', '3B', 'HR',
       'RBI', 'SB', 'CS', 'BB', 'SO', 'BA', 'OBP', 'SLG', 'OPS', 'OPS+', 'TB',
       'GDP', 'HBP', 'SH', 'SF', 'IBB'],
      dtype='object')

 The type of bos2017.columns: <class 'pandas.core.indexes.base.Index'>
Index(['Rk', 'Pos', 'Name', 'Age', 'G', 'PA', 'AB', 'R', 'H', '2B', '3B', 'HR',
       'RBI', 'SB', 'CS', 'BB', 'SO', 'BA', 'OBP', 'SLG', 'OPS', 'OPS+', 'TB',
       'GDP', 'HBP', 'SH', 'SF', 'IBB'],
      dtype='object')

 The type of bos2017.keys(): <class 'pandas.core.indexes.base.Index'>


### In-Class Exercise #2 - Describe & Sort Boston Red Sox Hitting Data <br>
<p>Take the data that you read in earlier from the Red Sox csv's and use the describe method to understand the data better. Compare the two years and decide which team is having the better year. Then sort the values based on Batting Average.</p>

In [142]:
# Setting display to show numbers in data frames to just 2 decimal places
pd.options.display.float_format = '{:,.2f}'.format

# Generating describe tables with only columns that have numbers
bos17stats = bos2017.describe(include = [np.number])
bos18stats = bos2018.describe(include = [np.number])

# Grabbing the stats I want from describe method and columns that I want from data frame
bos1 = bos17stats.loc[['mean', 'std', 'min', 'max']][['R', 'H', 'HR', 'RBI', 'BA', 'OBP', 'SLG', 'OPS+', 'TB']]
bos1

R      H    HR    RBI   BA  OBP  SLG   OPS+     TB
mean  34.13  63.43  7.30  31.96 0.28 0.35 0.39  93.52 100.09
std   30.80  55.65  8.31  32.03 0.16 0.15 0.16  78.69  91.93
min    0.00   1.00  0.00   0.00 0.11 0.11 0.20 -16.00   1.00
max  101.00 166.00 24.00 102.00 1.00 1.00 1.00 428.00 288.00

In [143]:
bos2 = bos18stats.loc[['mean', 'std', 'min', 'max']][['R', 'H', 'HR', 'RBI', 'BA', 'OBP', 'SLG', 'OPS+', 'TB']]
bos2

# Based on the two tables most significant statistics (R, H, HR, RBI, BA, OPS, TB), the 2018 Red Sox team
# has higher averages across almost all categories as well as higher peaks for those same stats

R      H    HR    RBI   BA  OBP  SLG   OPS+     TB
mean  43.60  75.25 10.40  41.30 0.24 0.31 0.39  86.45 127.20
std   38.29  62.46 11.72  37.56 0.06 0.07 0.14  52.20 113.23
min    0.00   1.00  0.00   0.00 0.09 0.14 0.09 -17.00   1.00
max  129.00 188.00 43.00 130.00 0.35 0.44 0.64 186.00 358.00

In [111]:
# Setting all DataFrames displayed to show all columns
pd.set_option('display.max_columns', None)

# Sorting table by Batting Average 
sorted_bos17 = bos2017.sort_values('BA', kind = 'mergesort', ascending = False)
sorted_bos17

Rk Pos                Name  Age    G   PA   AB    R    H  2B  3B  HR  RBI  \
22  23  2B      Chase d'Arnaud   30    2    1    1    2    1   0   0   0    0   
12  13  2B       Eduardo Nunez   30   38  173  165   23   53  12   0   8   27   
2    3  2B      Dustin Pedroia   33  105  463  406   46  119  19   0   7   62   
0    1   C   Christian Vazquez   26   99  345  324   43   94  18   2   5   32   
4    5  3B       Rafael Devers   20   58  240  222   34   63  14   0  10   30   
18  19  IF     Marco Hernandez   24   21   60   58    7   16   3   0   0    2   
3    4  SS     Xander Bogaerts   24  148  635  571   94  156  32   6  10   62   
5    6  LF   Andrew Benintendi   22  151  658  573   84  155  26   1  20   90   
17  18  UT         Tzu-Wei Lin   23   25   66   56    7   15   0   2   0    2   
7    8  RF        Mookie Betts   24  153  712  628  101  166  46   2  24  102   
16  17  UT          Sam Travis   23   33   83   76   13   20   6   0   0    1   
19  20  UT         Rajai Davis   36   17   38   36    7    9   2   0   0    2   
1    2  1B      Mitch Moreland   31  149  576  508   73  125  34   0  22   79   
6    7  CF  Jackie Bradley Jr.   27  133  541  482   58  118  19   3  17   63   
8    9  DH      Hanley Ramirez   33  133  553  496   58  120  24   0  23   62   
10  11  UT         Chris Young   33   90  276  243   30   57  12   2   7   25   
9   10   C          Sandy Leon   28   85  301  271   32   61  14   0   7   39   
14  15  IF       Josh Rutledge   28   37  118  107   10   24   2   1   0    9   
15  16  3B      Pablo Sandoval   30   32  108   99   10   21   2   0   4   12   
11  12  3B       Deven Marrero   26   71  188  171   32   36   9   0   4   27   
13  14  2B          Brock Holt   29   64  164  140   20   28   6   0   0    7   
21  22  UT       Blake Swihart   25    6    7    5    1    1   0   0   0    0   
20  21  UT        Steve Selsky   27    8    9    9    0    1   1   0   0    0   

    SB  CS  BB   SO     BA    OBP    SLG    OPS  OPS+   TB  GDP  HBP  SH  SF  \
22   0   0   0    0  1.000  1.000  1.000  2.000   428    1    0    0   0   0   
12   6   2   6   25  0.321  0.353  0.539  0.892   128   89    3    2   0   0   
2    4   3  49   48  0.293  0.369  0.392  0.760   100  159   11    2   2   4   
0    7   2  17   64  0.290  0.330  0.404  0.735    91  131   14    3   0   1   
4    3   1  18   57  0.284  0.338  0.482  0.819   111  107    5    0   0   0   
18   0   1   1   15  0.276  0.300  0.328  0.628    65   19    0    1   0   0   
3   15   1  56  116  0.273  0.343  0.403  0.746    95  230   17    6   0   2   
5   20   5  70  112  0.271  0.352  0.424  0.776   102  243   16    6   1   8   
17   1   1   9   17  0.268  0.369  0.339  0.709    88   19    0    0   1   0   
7   26   3  77   79  0.264  0.344  0.459  0.803   108  288    9    2   0   5   
16   1   0   6   23  0.263  0.325  0.342  0.667    75   26    2    1   0   0   
19   3   1   1   13  0.250  0.289  0.306  0.595    56   11    2    1   0   0   
1    0   1  57  120  0.246  0.326  0.443  0.769    99  225   14    6   0   5   
6    8   3  48  124  0.245  0.323  0.402  0.726    89  194    8    9   0   2   
8    1   3  51  116  0.242  0.320  0.429  0.750    94  213   15    6   0   0   
10   3   2  30   55  0.235  0.322  0.387  0.709    85   94    4    2   0   1   
9    0   0  25   74  0.225  0.290  0.354  0.644    68   96    5    1   1   3   
14   1   0   9   31  0.224  0.297  0.262  0.558    49   28    1    2   0   0   
15   0   1   8   24  0.212  0.269  0.354  0.622    61   35    4    0   0   1   
11   5   0  12   61  0.211  0.259  0.333  0.593    54   57    8    0   3   2   
13   2   1  19   34  0.200  0.305  0.243  0.548    47   34    3    3   0   2   
21   0   0   2    3  0.200  0.429  0.200  0.629    74    1    0    0   0   0   
20   0   0   0    5  0.111  0.111  0.222  0.333   -16    2    0    0   0   0   

    IBB  
22    0  
12    0  
2     4  
0     0  
4     3  
18    0  
3     6  
5     7  
17    0  
7     9  
16    0  
19    0  
1     6 

In [88]:
# Sorting table by Batting Average 
sorted_bos18 = bos2018.sort_values('BA', kind = 'mergesort', ascending = False)
sorted_bos18

Rk Pos                Name  Age    G   PA   AB    R    H  2B  3B  HR  RBI  \
7    8  RF        Mookie Betts   25  136  614  520  129  180  47   5  32   80   
8    9  DH       J.D. Martinez   30  150  649  569  111  188  37   2  43  130   
5    6  LF   Andrew Benintendi   23  148  661  579  103  168  41   6  16   87   
3    4  SS     Xander Bogaerts   25  136  580  513   72  148  45   3  23  103   
13  14  1B        Steve Pearce   35   50  165  136   19   38   8   1   7   26   
9   10  MI          Brock Holt   30  109  367  321   41   89  18   2   7   46   
2    3  2B       Eduardo Nunez   31  127  502  480   56  127  23   3  10   44   
12  13  UT      Hanley Ramirez   34   44  195  177   25   45   7   0   6   29   
15  16  SS         Tzu-Wei Lin   24   37   73   65   15   16   6   1   1    6   
1    2  1B      Mitch Moreland   32  124  459  404   57   99  23   4  15   68   
14  15  2B         Ian Kinsler   36   37  143  132   17   32   6   0   1   16   
4    5  3B       Rafael Devers   21  121  490  450   59  108  24   0  21   66   
6    7  CF  Jackie Bradley Jr.   28  144  535  474   76  111  33   4  13   59   
11  12  UT       Blake Swihart   26   82  207  192   28   44  10   0   3   18   
16  17  UT          Sam Travis   24   19   38   36    5    8   3   0   1    7   
10  11   C   Christian Vazquez   27   80  269  251   24   52  10   0   3   16   
0    1   C          Sandy Leon   29   89  288  265   30   47  12   0   5   22   
19  20   C          Dan Butler   31    2    7    6    0    1   0   0   0    1   
17  18  UT    Brandon Phillips   37    9   27   23    4    3   0   0   1    2   
18  19  2B      Dustin Pedroia   34    3   13   11    1    1   0   0   0    0   

    SB  CS  BB   SO     BA    OBP    SLG    OPS  OPS+   TB  GDP  HBP  SH  SF  \
7   30   6  81   91  0.346  0.438  0.640  1.078   186  333    5    8   0   5   
8    6   1  69  146  0.330  0.402  0.629  1.031   173  358   19    4   0   7   
5   21   3  71  106  0.290  0.366  0.465  0.830   123  269    9    2   2   7   
3    8   2  55  102  0.288  0.360  0.522  0.883   135  268   14    6   0   6   
13   0   0  22   27  0.279  0.394  0.507  0.901   141   69    1    5   0   2   
9    7   7  37   73  0.277  0.362  0.411  0.774   109  132    7    7   0   2   
2    7   2  16   69  0.265  0.289  0.388  0.677    81  186   17    2   1   3   
12   4   1  14   35  0.254  0.313  0.395  0.708    90   70    9    2   0   2   
15   0   1   8   17  0.246  0.329  0.415  0.744   100   27    0    0   0   0   
1    2   0  50  102  0.245  0.325  0.433  0.758   102  175   12    0   0   5   
14   7   3  10   24  0.242  0.294  0.311  0.604    64   41    5    0   0   1   
4    5   2  38  121  0.240  0.298  0.433  0.731    94  195    9    0   0   2   
6   17   1  46  137  0.234  0.314  0.403  0.717    92  191    6   11   0   4   
11   6   1  15   57  0.229  0.285  0.328  0.613    65   63    4    0   0   0   
16   0   0   2   10  0.222  0.263  0.389  0.652    73   14    1    0   0   0   
10   4   1  13   41  0.207  0.257  0.283  0.540    46   71    5    4   1   0   
0    1   0  15   75  0.177  0.232  0.279  0.511    37   74    6    4   3   1   
19   0   0   0    2  0.167  0.143  0.167  0.310   -17    1    0    0   0   1   
17   0   0   4    7  0.130  0.259  0.261  0.520    42    6    1    0   0   0   
18   0   0   2    1  0.091  0.231  0.091  0.322    -7    1    0    0   0   0   

    IBB  
7     8  
8    11  
5     1  
3     4  
13    0  
9     2  
2     0  
12    0  
15    0  
1     2  
14    0  
4     6  
6     3  
11    0  
16    0  
10    1  
0     0  
19    0  
17    0  
18    0

### Filtration <br>
<p>Let's look at how to filter dataframes for rows that fulfill a specific conditon.</p>

##### Conditionals

In [144]:
# Conditional boolean dataframe

data['ages'] >= 25


0     True
1    False
2    False
3    False
4     True
5    False
6    False
7     True
8    False
9    False
Name: ages, dtype: bool

##### Subsetting

In [145]:
# exactly like numpy

data[data['ages'] >= 25]


names  ages
0   Alice    25
4    John    29
7  Lauren    28

### Column Transformations <br>
<p>Rarely, if ever, will the columns in the original raw dataframe read from CSV or database table be the ones you actually need for your analysis. You will spend lots of time constantly transforming columns or groups of columns using general computational operations to produce new ones that are functions of the old ones. Pandas has full support for this: Consider the following dataframe containing membership term and renewal number for a group of customers:</p>

In [148]:
# Generate some fake data
customer_id = np.random.randint(1000,1100,10)
renewal_nbr = np.random.randint(0,10,10)
customer_dict = {1: 0.5, 0: 1}
terms_in_years = [customer_dict[key] for key in np.random.randint(0,2,10)]

random_data = {
    'customer_id': customer_id,
    'renewal_nbr': renewal_nbr,
    'terms_in_years': terms_in_years
}

customers = pd.DataFrame.from_dict(random_data)
customers

customer_id  renewal_nbr  terms_in_years
0         1047            4            0.50
1         1037            0            1.00
2         1047            8            1.00
3         1089            0            0.50
4         1035            7            1.00
5         1039            7            1.00
6         1041            7            1.00
7         1004            5            1.00
8         1012            8            1.00
9         1010            1            0.50

##### Generating a New Column w/Data

In [159]:
# DataFrame['key'] = Some Calculation from our DataFrame Columns

customers['customer_tenure'] = customers['renewal_nbr']*customers['terms_in_years']
customers

customer_id  renewal_nbr  terms_in_years  customer_tenure  ages  \
0         1047            4            0.50             2.00    22   
1         1037            0            1.00             0.00    28   
2         1047            8            1.00             8.00    43   
3         1089            0            0.50             0.00    50   
4         1035            7            1.00             7.00    36   
5         1039            7            1.00             7.00    59   
6         1041            7            1.00             7.00    34   
7         1004            5            1.00             5.00    64   
8         1012            8            1.00             8.00    49   
9         1010            1            0.50             0.50    24   

     age_group  loyalty_age_group  
0  Young Adult    New Young Adult  
1  Young Adult    New Young Adult  
2        Adult        Loyal Adult  
3        Adult          New Adult  
4        Adult        Loyal Adult  
5        Adult        Loyal Adult  
6  Young Adult  Loyal Young Adult  
7        Adult        Loyal Adult  
8        Adult        Loyal Adult  
9  Young Adult    New Young Adult

##### User Defined Function

<p>If what you want to do to a column that can't be represented by simple mathematical operations, you can write your own $\textit{user defined function}$ with the full customizability available in Python and any external Python packages, then map it directly onto a column. Let's add some ages to our customer dataframe, and then classify them into our custom defined grouping scheme:</p>

In [197]:
# use .apply to map over dataframe

# Create a new column for ages
#customers['ages'] = np.random.randint(18,70,10)

# User defined function
def make_age_groups(age):
    if 10 <= age and 20 > age:
        return 'Teenager'
    elif 20 <= age and 35 > age:
        return 'Young Adult'
    elif 35 <= age and 65 > age:
        return 'Adult'
    elif 65 <= age:
        return 'Senior'


# print(make_age_groups)
customers['age_group'] = customers['ages'].apply(make_age_groups)
customers


customer_id  renewal_nbr  terms_in_years  customer_tenure  ages  \
0         1047            4           0.500            2.000    22   
1         1037            0           1.000            0.000    28   
2         1047            8           1.000            8.000    43   
3         1089            0           0.500            0.000    50   
4         1035            7           1.000            7.000    36   
5         1039            7           1.000            7.000    59   
6         1041            7           1.000            7.000    34   
7         1004            5           1.000            5.000    64   
8         1012            8           1.000            8.000    49   
9         1010            1           0.500            0.500    24   

     age_group  loyalty_age_group  
0  Young Adult    New Young Adult  
1  Young Adult    New Young Adult  
2        Adult        Loyal Adult  
3        Adult          New Adult  
4        Adult        Loyal Adult  
5        Adult        Loyal Adult  
6  Young Adult  Loyal Young Adult  
7        Adult        Loyal Adult  
8        Adult        Loyal Adult  
9  Young Adult    New Young Adult

<p>As a last example I'll show here how you would use a lambda function to create a UDF that depends on $\textit{more than one}$ column:</p>

<li>UDF = User Defined Function</li>

In [155]:
#Axis for apply can only be 1 or 0 -- 1 being the X axis 0 being the Y axis
def make_loyalty_age_group(row):
    age = row['ages']
    tenure = row['customer_tenure']
    
    if 10 <= age and 20 > age:
        age_group = 'Teenager'
    elif 20 <= age and 35 > age:
        age_group = 'Young Adult'
    elif 35 <= age and 65 > age:
        age_group = 'Adult'
    elif 65 <= age:
        age_group = 'Senior'
    
    if tenure > 2.0:
        loyalty_age_group = f'Loyal {age_group}'
    else:
        loyalty_age_group = f'New {age_group}'
    
    return loyalty_age_group

customers['loyalty_age_group'] = customers.apply(make_loyalty_age_group, axis = 1)
customers


customer_id  renewal_nbr  terms_in_years  customer_tenure  ages  \
0         1047            4            0.50             2.00    22   
1         1037            0            1.00             0.00    28   
2         1047            8            1.00             8.00    43   
3         1089            0            0.50             0.00    50   
4         1035            7            1.00             7.00    36   
5         1039            7            1.00             7.00    59   
6         1041            7            1.00             7.00    34   
7         1004            5            1.00             5.00    64   
8         1012            8            1.00             8.00    49   
9         1010            1            0.50             0.50    24   

     age_group  loyalty_age_group  
0  Young Adult    New Young Adult  
1  Young Adult    New Young Adult  
2        Adult        Loyal Adult  
3        Adult          New Adult  
4        Adult        Loyal Adult  
5        Adult        Loyal Adult  
6  Young Adult  Loyal Young Adult  
7        Adult        Loyal Adult  
8        Adult        Loyal Adult  
9  Young Adult    New Young Adult

### In-Class Exercise #3 - Create Your Own UDF <br>
<p>Using the Boston Red Sox data, create your own UDF which creates a new column called 'All-Star' and puts every player with either a batting average over .280 or an on base percentage of over .360 with a result of 'Yes' in the column and 'No' if not.</p>

In [188]:
pd.options.display.float_format = '{:,.3f}'.format
"""
    Name  BA OBP AllStar
    --------------------
    Name .233 .360 Yes
    Name .150 .288 No
"""

def make_all_star(row):
    BA = row['BA']
    OBP = row['OBP']
    group = ''
    
    if BA > 0.28 or OBP > 0.36:
        group = 'AllStar'
    
    if group:
        AllStar = 'Yes'
    else:
        AllStar = 'No'
    
    return AllStar

# Apply UDF and show only relevant columns in DataFrame using iloc
bos2017['AllStar'] = bos2017.apply(make_all_star, axis = 1)
bos2017.iloc[:, [2,17,18,28]]

Name    BA   OBP AllStar
0    Christian Vazquez 0.290 0.330     Yes
1       Mitch Moreland 0.246 0.326      No
2       Dustin Pedroia 0.293 0.369     Yes
3      Xander Bogaerts 0.273 0.343      No
4        Rafael Devers 0.284 0.338     Yes
5    Andrew Benintendi 0.271 0.352      No
6   Jackie Bradley Jr. 0.245 0.323      No
7         Mookie Betts 0.264 0.344      No
8       Hanley Ramirez 0.242 0.320      No
9           Sandy Leon 0.225 0.290      No
10         Chris Young 0.235 0.322      No
11       Deven Marrero 0.211 0.259      No
12       Eduardo Nunez 0.321 0.353     Yes
13          Brock Holt 0.200 0.305      No
14       Josh Rutledge 0.224 0.297      No
15      Pablo Sandoval 0.212 0.269      No
16          Sam Travis 0.263 0.325      No
17         Tzu-Wei Lin 0.268 0.369     Yes
18     Marco Hernandez 0.276 0.300      No
19         Rajai Davis 0.250 0.289      No
20        Steve Selsky 0.111 0.111      No
21       Blake Swihart 0.200 0.429     Yes
22      Chase d'Arnaud 1.000 1.000     Yes

### Aggregations <br>
<p>The raw data plus some transformations is generally only half the story. Your objective is to extract actual insights and actionable conclusions from the data, and that means reducing it from potentially billions of rows to some summary statistics via aggregation functions.</p>

##### groupby() <br>
<p>The .groupby() function is in some ways a 'master' aggregation.</p> 

<p>Data tables will usually reserve one column as a primary key - that is, a column for which each row has a unique value. This is to facilitate access to the exact rows of a data table that a user wants to view. The other columns will often have repeated values, such as the age groups in the above examples. We can use these columns to explore the data using the Pandas API:</p>

In [190]:
# also introducing .count() here, exact same as to how it's used in SQL

# Using the groupby with the column intact as a column/key
# customers.groupby('age_group', as_index = False).count()[['customer_id','age_group']]

customers.groupby('age_group').count()
print(type(customers.groupby('age_group')))







<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


##### Type of groupby()

<p>The result is a new dataframe, the columns of which all contain the counts of the grouped field. Notice the type of a grouped dataframe:</p>

In [192]:
print(type(customers), 'Regular Data Frame')
print(f'\n Type of Group By for Customer.GroupBy: {type(customers.groupby("age_group"))}')

<class 'pandas.core.frame.DataFrame'> Regular Data Frame

 Type of Group By for Customer.GroupBy: <class 'pandas.core.groupby.generic.DataFrameGroupBy'>


<p>This is because simply grouping data doesn't quite make sense without an aggregation function like count() to pair with. In this case, we're counting occurances of the grouped field, but that's not all we can do. We can take averages, standard deviations, mins, maxes and much more! Let's see how this works a bit more:</p>

##### mean()

In [222]:
# mean = average

customers.groupby('age_group').mean(numeric_only = True)[['customer_tenure', 'ages']]


customer_tenure   ages
age_group                          
Adult                  5.833 50.167
Young Adult            2.375 27.000

##### groupby() w/Multiple Columns

<p>We end up with the average age of the groups in the last column, the average tenure in the tenure column, and so on and so forth. You can even split the groups more finely by passing a list of columns to group by:</p>

In [210]:
customers.groupby(['age_group', 'ages']).mean(numeric_only = True).sort_values('ages', ascending = False)

customer_id  renewal_nbr  terms_in_years  customer_tenure
age_group   ages                                                           
Adult       64      1,004.000        5.000           1.000            5.000
            59      1,039.000        7.000           1.000            7.000
            50      1,089.000        0.000           0.500            0.000
            49      1,012.000        8.000           1.000            8.000
            43      1,047.000        8.000           1.000            8.000
            36      1,035.000        7.000           1.000            7.000
Young Adult 34      1,041.000        7.000           1.000            7.000
            28      1,037.000        0.000           1.000            0.000
            24      1,010.000        1.000           0.500            0.500
            22      1,047.000        4.000           0.500            2.000

##### drop_duplicates()

<p>Drops all duplicates from the current dataframe</p>

In [211]:
customers

customer_id  renewal_nbr  terms_in_years  customer_tenure  ages  \
0         1047            4           0.500            2.000    22   
1         1037            0           1.000            0.000    28   
2         1047            8           1.000            8.000    43   
3         1089            0           0.500            0.000    50   
4         1035            7           1.000            7.000    36   
5         1039            7           1.000            7.000    59   
6         1041            7           1.000            7.000    34   
7         1004            5           1.000            5.000    64   
8         1012            8           1.000            8.000    49   
9         1010            1           0.500            0.500    24   

     age_group  loyalty_age_group  
0  Young Adult    New Young Adult  
1  Young Adult    New Young Adult  
2        Adult        Loyal Adult  
3        Adult          New Adult  
4        Adult        Loyal Adult  
5        Adult        Loyal Adult  
6  Young Adult  Loyal Young Adult  
7        Adult        Loyal Adult  
8        Adult        Loyal Adult  
9  Young Adult    New Young Adult

In [213]:
customers_copy = customers.drop_duplicates('renewal_nbr').reset_index(drop=True)
customers_copy

customer_id  renewal_nbr  terms_in_years  customer_tenure  ages  \
0         1047            4           0.500            2.000    22   
1         1037            0           1.000            0.000    28   
2         1047            8           1.000            8.000    43   
3         1035            7           1.000            7.000    36   
4         1004            5           1.000            5.000    64   
5         1010            1           0.500            0.500    24   

     age_group loyalty_age_group  
0  Young Adult   New Young Adult  
1  Young Adult   New Young Adult  
2        Adult       Loyal Adult  
3        Adult       Loyal Adult  
4        Adult       Loyal Adult  
5  Young Adult   New Young Adult

In [214]:
# Send customer data into a CSV file
customers_copy.to_csv('files/customers.csv')

<p>Thus the groupby operation allows you to rapidly make summary observations about the state of your entire dataset at flexible granularity. In one line above, we actually did something very complicated - that's the power of the dataframe. In fact, the process often consists of several iterative groupby operations, each revealing greater insight than the last - if you don't know where to start with a dataset, try a bunch of groupbys!</p>

### Homework Excersise #1 - Find the Total Number of Runs and RBIs for the Red Sox <br>
<p>Get total number of home runs and rbi's</p>

In [240]:
# step 1: Add a new column with the key 'Team' and all column values should be 'BOS'

# step 2: Group by the 'Team' column and get total home runs and rbi's

# Produce data for both 2017 and 2018 (ie print both seperated by a newline character \n)

"""
TEAM    HR   RBI
----------------
BOS     144  538
"""

bos2017['Team'] = 'Boston'
bos2018['Team'] = 'Boston'
print('Red Sox 2017 Stats: \n\n',bos2017.groupby('Team').sum(numeric_only = True)[['HR', 'RBI']])
print('\nRed Sox 2018 Stats: \n\n',bos2018.groupby('Team').sum(numeric_only = True)[['HR', 'RBI']])

Red Sox 2017 Stats: 

          HR  RBI
Team            
Boston  168  735

Red Sox 2018 Stats: 

          HR  RBI
Team            
Boston  208  826


## Find the average age of runners in the 2017 Boston Marathon

In [251]:
pd.options.display.float_format = '{:,.2f}'.format

# The average age of male and female runners
marathon.groupby('M/F').mean(numeric_only=True)[['Age']]

Age
M/F      
F   39.95
M   44.77

In [252]:
# The average age of all runners
print("The average age of all runners for the 2017 marathon is: ", marathon.mean(numeric_only=True)[['Age']])

The average age of all runners for the 2017 marathon is:  Age   42.59
dtype: float64
